File: src/splprime/scan/Scanner.java
package splprime.scan;

import java.util.ArrayList;
import java.util.List;

import java.util.Optional;

public class Scanner {

	// In and output
	private final String source;
	private final List<Token> tokens = new ArrayList<>();
	private int current = 0;
	private int start = 0;
	private int line = 1;

	private boolean atEnd() {
		return current == source.length();
	}

	private char advance() {
		return source.charAt(current++);
	}

	private boolean match(char expected) {
		if (atEnd())
			return false;

		if (source.charAt(current) != expected)
			return false;

		current++;
		return true;
	}

	private Optional<Character> peek() {
		if (atEnd()) {
			return Optional.empty();
		}

		return Optional.of(source.charAt(current));

	}

	public Scanner(String source) {
		this.source = source;
	}

	private String currentLexeme() {
		return source.substring(start, current);
	}

	private void addToken(TokenType type) {
		Token tok = new Token(type, currentLexeme(), type, line);
		tokens.add(tok);
	}

	private void addToken(TokenType type, String lexeme) {
		Token tok = new Token(type, lexeme, type, line);
		tokens.add(tok);
	}

	/*
	 * private TokenType resolveSingleCharacterTokenType(char c) {
	 * switch (c) {
	 * case '+':
	 * return TokenType.OP_PLUS;
	 * case '-':
	 * return TokenType.OP_MINUS;
	 * case '*':
	 * return TokenType.OP_STAR;
	 * case ';':
	 * return TokenType.CHAR_SEMICOLON;
	 * case '(':
	 * return TokenType.CHAR_L_PAREN;
	 * case ')':
	 * return TokenType.CHAR_R_PAREN;
	 * case '{':
	 * return TokenType.CHAR_L_BRACE;
	 * case '}':
	 * return TokenType.CHAR_R_BRACE;
	 * 
	 * default:
	 * throw new RuntimeException("invalid single character token: " + c);
	 * }
	 * }
	 */
	private void handleNumber(char c) {
		boolean hasDecimal = false;
		boolean hasAfterDecimal = false;
		boolean done = false;

		while (!done) {
			char next = peek().orElse('\n');
			switch (next) {
				case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
					advance();
					hasAfterDecimal = hasDecimal;

					break;
				case '.':
					if (hasDecimal) {
						throw new ScannerException(line, "number error: expected at most one decimal point.");
					}

					hasDecimal = true;
					advance();

					break;
				default:
					done = true;
					if (hasDecimal && !hasAfterDecimal) {
						throw new ScannerException(line, "number error: expected numbers after decimal point.");
					}

					break;
			}
		}

		addToken(TokenType.LIT_NUMBER);
	}

	private static boolean isAlphaNumeric(Optional<Character> c) {
		if (c.isEmpty()) {
			return false;
		}

		char chr = c.get();
		return Character.isAlphabetic(chr) || chr >= '0' && chr <= '9';
	}

	private void handleAlpha(char c) {
		while (isAlphaNumeric(peek()))
			advance();

		String text = currentLexeme();

		var type = TokenResolver.resolveKeywordOrIdent(text);
		addToken(type);
	}

	private void handleString(char c) {
		while (peek().orElse('\n') != '"')
			advance();

		if (!match('"'))
			throw new ScannerException(line, "lexer error: unterminated string");

		// remove quotes from lexeme
		String text = currentLexeme();
		text = text.substring(1, text.length() - 1);

		addToken(TokenType.LIT_STRING, text);
	}

	// Scan tokens
	public List<Token> scanTokens() {

		while (!atEnd()) {
			start = current;
			char c = advance();
			TokenType type;

			switch (c) {
				case '\n':
					line++;
					// throw new RuntimeException();
					break;
				case '+', '-', '*', ';', '(', ')', '{', '}', '!':
					type = TokenResolver.resolveStatic(String.valueOf(c));
					addToken(type);
					break;
				case '/':
					if (match('/')) {
						while (peek().orElse('\n') != '\n')
							advance();

					} else {
						addToken(TokenType.OP_SLASH);
					}
					break;
				case '=', '<', '>':
					var tokStr = String.valueOf(c);

					if (match('=')) {
						tokStr += '=';
					}
					addToken(TokenResolver.resolveStatic(tokStr));
					break;
				case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
					handleNumber(c);
					break;
				case '"':
					handleString(c);
					break;
				default:
					if (Character.isAlphabetic(c)) {
						handleAlpha(c);
						continue;
					}
					if (Character.isWhitespace(c)) {
						continue;
					}

					throw new ScannerException(line, "lexical error: unexpected token: " + c);
			}

		}
		start = current;
		addToken(TokenType.EOF);

		return tokens;
	}

}

File: pkg/splprime/scan/scanner.go
package scan

import (
	"errors"
	"strconv"
	"unicode"

	"github.com/remogeissbuehler/compiler-construction/pkg/option"
)

func isAlphaNumeric(r option.Option[rune]) bool {
	return r.IsSomeAnd(func(x rune) bool {
		return unicode.IsLetter(x) || unicode.IsDigit(x)
	})
}

type Scanner struct {
	source  []rune
	tokens  []Token
	current int
	start   int
	line    int
}

func NewScanner(source string) Scanner {
	return Scanner{
		source:  []rune(source),
		tokens:  make([]Token, 0, 64),
		current: 0,
		start:   0,
		line:    1,
	}
}

func (s Scanner) atEnd() bool {
	return s.current == len(s.source)
}

func (s *Scanner) advance() rune {
	curr := s.current
	s.current++
	return s.source[curr]
}

func (s *Scanner) match(expected rune) bool {
	if s.atEnd() {
		return false
	}

	if s.source[s.current] != expected {
		return false
	}

	s.current++
	return true
}

func (s *Scanner) peek() option.Option[rune] {
	if s.atEnd() {
		return option.None[rune]()
	}

	return option.Some(s.source[s.current])
}

func (s *Scanner) currentLexeme() string {
	return string(s.source[s.start:s.current])
}

func (s *Scanner) addToken(typ TokenType) {
	s.addTokenWithLiteral(typ, nil)
}

func (s *Scanner) addTokenWithLiteral(typ TokenType, literal any) {
	tok := NewToken(typ, s.currentLexeme(), literal, s.line)
	s.tokens = append(s.tokens, tok)
}

func (s *Scanner) handleNumber() error {
	for s.peek().IsSomeAnd(unicode.IsDigit) {
		s.advance()
	}

	if s.match('.') {
		for s.peek().IsSomeAnd(unicode.IsDigit) {
			s.advance()
		}
	}

	val, err := strconv.ParseFloat(s.currentLexeme(), 64)
	if err != nil {
		return errors.New("number error: cannot parse float from number")
	}

	s.addTokenWithLiteral(LitNumber, val)
	return nil
}

func (s *Scanner) handleAlpha() {
	for isAlphaNumeric(s.peek()) {
		s.advance()
	}

	text := s.currentLexeme()
	typ := ResolveKeywordOrIdent(text)
	s.addToken(typ)
}

func (s *Scanner) handleString() error {
	var n rune

	for {
		n = s.peek().OrElse('\n')
		if n == '"' || n == '\n' {
			break
		}

		s.advance()
	}

	if !s.match('"') {
		return errors.New("lexer error: unterminated string")
	}

	text := s.currentLexeme()
	text = text[1 : len(text)-1]

	s.addTokenWithLiteral(LitString, text)
	return nil
}

func (s *Scanner) ScanTokens() ([]Token, error) {
	for !s.atEnd() {
		s.start = s.current
		r := s.advance()

		switch r {

		case '\n':
			s.line++
		case '+', '-', '*', ';', '(', ')', '{', '}', '!':
			typ, err := ResolveStaticToken(string(r))
			if err != nil {
				return nil, err
			}
			s.addToken(typ)
		case '/':
			if s.match('/') {
				for s.peek().OrElse('\n') != '\n' {
					s.advance()
				}
			} else {
				s.addToken(OpSlash)
			}
		case '=', '<', '>':
			tokStr := string(r)

			if s.match('=') {
				tokStr += "="
			}

			typ, err := ResolveStaticToken(tokStr)
			if err != nil {
				return nil, err
			}
			s.addToken(typ)
		case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
			s.handleNumber()
		case '"':
			s.handleString()
		default:
			if unicode.IsLetter(r) {
				s.handleAlpha()
				continue
			}

			if unicode.IsSpace(r) {
				continue
			}

			return nil, errors.New("lexical error: unexpected token: " + string(r))
		}
	}
	s.start = s.current
	s.addToken(EOF)

	return s.tokens, nil
}
